src/sql-parser/src/lexer.rs

```diff
@@ -0,0 +1,358 @@
+    buf.take_while(|ch| ch != '\n');
+}
+
+fn lex_multiline_comment(buf: &mut LexBuf) -> Result<(), ParserError> {
+    let pos = buf.pos() - 2;
+    let mut nesting = 0;
+    while let Some(ch) = buf.next() {
+        match ch {
+            '*' if buf.consume('/') => {
+                if nesting == 0 {
+                    return Ok(());
+                } else {
+                    nesting -= 1;
+                }
+            }
+            '/' if buf.consume('*') => nesting += 1,
+            _ => (),
+        }
+    }
+    Err(buf.err(pos, "unterminated multiline comment"))
+}
+
+fn lex_ident(buf: &mut LexBuf) -> Token {
+    buf.prev();
+    let word = buf.take_while(
+        |ch| matches!(ch, 'A'..='Z' | 'a'..='z' | '\u{80}'..='\u{ff}' | '0'..='9' | '$' | '_'),
+    );
+    let word_uppercase = word.to_uppercase();
+    Token::Word(Word {
+        value: word,
+        quoted: false,
+        keyword: if ALL_KEYWORDS.contains(&word_uppercase.as_str()) {
+            word_uppercase
+        } else {
+            "".into()
+        },
+    })
+}
+
+fn lex_quoted_ident(buf: &mut LexBuf) -> Result<Token, ParserError> {
+    let mut s = String::new();
+    let pos = buf.pos() - 1;
+    loop {
+        match buf.next() {
+            Some('"') if buf.consume('"') => s.push('"'),
+            Some('"') => break,
+            Some(c) => s.push(c),
+            None => return Err(buf.err(pos, "unterminated quoted identifier")),
+        }
+    }
+    Ok(Token::Word(Word {
+        value: s,
+        quoted: true,
+        keyword: "".into(),
+    }))
+}
+
+fn lex_string(buf: &mut LexBuf) -> Result<String, ParserError> {
+    let mut s = String::new();
+    loop {
+        let pos = buf.pos() - 1;
+        loop {
+            match buf.next() {
+                Some('\'') if buf.consume('\'') => s.push('\''),
+                Some('\'') => break,
+                Some(c) => s.push(c),
+                None => return Err(buf.err(pos, "unterminated quoted string")),
+            }
+        }
+        // Concatenate string literals that are separated by whitespace if and
+        // only if that whitespace contains at least one newline character. This
+        // bizzare rule matches PostgreSQL and the SQL standard.
+        let whitespace = buf.take_while(|ch| matches!(ch, ' ' | '\t' | '\n' | '\r' | '\x0c'));
+        if !(whitespace.contains(&['\n', '\r'][..]) && buf.consume('\'')) {
+            return Ok(s);
+        }
+    }
+}
+
+fn lex_extended_string(buf: &mut LexBuf) -> Result<Token, ParserError> {
+    fn lex_unicode_escape(buf: &mut LexBuf, n: usize) -> Result<char, ParserError> {
+        let pos = buf.pos();
+        buf.next_n(n)
+            .and_then(|s| u32::from_str_radix(s, 16).ok())
+            .and_then(|codepoint| char::try_from(codepoint).ok())
+            .ok_or_else(|| buf.err(pos, "invalid unicode escape"))
+    }
+
+    let mut s = String::new();
+    let pos = buf.pos() - 1;
+    loop {
```

Comment:
Good catch. I had thought that wasn't allowed because you can't write

```
e'a'
e'b'
```

but that was my error. You can totally write:

```
e'a'
'b'
```